keys,sqlbase: move PublicSchemaID to sqlbase

ajwerner opened the pull request:
The PublicSchemaID was introduced during the work for temp tables and the
associated migration of the namespace table. The chosen constantly was
erroneously put into the `PseudoTableIDs` which is used to create split
points and control zone configs for low-level constructs like meta ranges,
liveness range, timeseries, etc.

This commit also fixed non-sense I introduced in #39638 while trying to
enable real system tables to have their own zone configs (as opposed to
just being controlled by the system zone config). All of the pseudo-table
ids had corresponding filters below.

This change should get backported as far back as 19.1.

Release note (bug fix): Fix bug where an extra split was created in the
keyspace which does not correspond to user data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47946)
<!-- Reviewable:end -->

ajwerner commented:
cc @jordanlewis for the `PublicSchemaID` and @andreimatei because we talked about this.

I want to add some testing about the splits.

My sense is 29 is a system table ID up for grabs. @lucy-zhang it's especially important to note the current usage of that 29 value.

andreimatei reviewed (commented):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47946#-:-M5by9x27uFb2FkZLf0j:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/keys/constants.go, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47946#-M5bxs0p8jO_ezTDlqHG:-M5bxs0p8jO_ezTDlqHH:b-wnn32k) ([raw file](https://github.com/cockroachdb/cockroach/blob/91db3e274bf984ce40d4b659ab4b0e6f8814e0d6/pkg/keys/constants.go#L356)):*
> ```Go
> 	ReplicationStatsTableID              = 27
> 	ReportsMetaTableID                   = 28
> 	PublicSchemaID                       = 29
> ```

do you think we'll end up writing descriptors for user-defined schemas (and non-user defined ones too) in the same namespace where we write table and database descriptors? If so - if it's likely we'll write a descriptor with ID 29 - I'd keep a note here about what 29 is.


<!-- Sent from Reviewable.io -->

ajwerner commented:
> do you think we'll end up writing descriptors for user-defined schemas (and non-user defined ones too) in the same namespace where we write table and database descriptors? If so - if it's likely we'll write a descriptor with ID 29 - I'd keep a note here about what 29 is.

I do think we will write schema descriptors into the same namespace. That being said, they'll have an ID per database so I don't think we'll ever use this 29. What database would it be the child of?

thoszhang reviewed (commented):
If we're not running low on system table IDs, can we just avoid reusing 29 anyway to avoid confusion?

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47946#-:-M5i2mSP-U1igjSGYufX:bbri517)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Yeah I'm thinking I'll leave an unexported const also as an anchor for commentary. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47946#-:-M5i5EVG3JF7KRfBIuqj:b-u1gmvj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Given that this also manifests as a possibly under-replicated range in a mixed version cluster, perhaps we could note as much in the release note? I'm unsure if this is the PR touching that bug. Internal thread [here](https://cockroachlabs.slack.com/archives/CARM0PTK4/p1590600800290500).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47946#-:-M8MFsTu7w1F8-Be8mPZ:by8oa1c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Commit:
keys,sqlbase: move PublicSchemaID to sqlbase, prevent splits on pseudo IDs

The PublicSchemaID was introduced during the work for temp tables and the
associated migration of the namespace table. The chosen constantly was
erroneously put into the PseudoTableIDs which is used to create split
points and control zone configs for low-level constructs like meta ranges,
liveness range, timeseries, etc.

This commit also fixed non-sense I introduced in #39638 while trying to
enable real system tables to have their own zone configs (as opposed to
just being controlled by the system zone config). All of the pseudo-table
ids had corresponding filters below.

This change should get backported as far back as 19.1.

Release note (bug fix): Fix bug where an extra split was created in the
keyspace which does not correspond to user data which could lead to a
report of an under-replicated range in mixed-version 19.2-20.1 clusters.

ajwerner commented:
Sorry I let this sit for so long. I've updated it.

ajwerner commented:
actually hold off, needs more

ajwerner commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/50915
